### PR TITLE
[WIP] Backport `prusti_assert!` and `prusti_assume!`

### PR DESCRIPTION
**Pull request by zgrannan**
This should enable `prusti_assert!` and `prusti_assume!` to work when the `unsafe_core_proof` option is not enabled.

### CODE REVIEW

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 1]
```diff
@@ -399,6 +524,7 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         self.cfg_method = self.encoder.patch_snapshots_method(self.cfg_method)
             .with_span(mir_span)?;
 
+
```
Comment (Aurel300) [inline prusti-viper/src/encoder/procedure_encoder.rs]: Remove the whitespace changes.

File: prusti-viper/src/encoder/specification_blocks.rs [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+pub(super) struct SpecificationBlocks {
+    /// All blocks that are generated as a result of specification type-checking.
+    specification_blocks: BTreeSet<mir::BasicBlock>,
+    /// Blocks through which specifications are entered.
+    specification_entry_blocks: BTreeSet<mir::BasicBlock>,
+    ghost_blocks: BTreeSet<mir::BasicBlock>,
+    /// A set of blocks containing the loop invariant of a given loop in
+    /// execution order.
+    ///
+    /// FIXME: Add a check that ensure that the blocks are one after another.
```
Comment (Aurel300) [inline prusti-viper/src/encoder/specification_blocks.rs]: Do you mean that all `body_invariant!(...)` are placed next to each other? If so, the proper fix is rather to encode invariants properly no matter where they are placed.

File: vir/defs/polymorphic/ast/stmt.rs [hunk 3]
```diff
@@ -167,6 +169,18 @@ impl fmt::Display for Assert {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+pub struct Assume {
```
Comment (Aurel300) [inline vir/defs/polymorphic/ast/stmt.rs]: Why not re-use `inhale` for the assumptions?

File: prusti-viper/src/encoder/specification_blocks.rs [hunk 4]
```diff
@@ -0,0 +1,174 @@
+use std::collections::{BTreeMap, BTreeSet};
```
Comment (Aurel300) [inline prusti-viper/src/encoder/specification_blocks.rs]: Is this file copied from something similar for the VIR high encoding? Can it use this directly?

File: prusti-viper/src/encoder/specification_blocks.rs (hunk 2 above)
Comment (zgrannan) [reply prusti-viper/src/encoder/specification_blocks.rs]: As we now use the `SpecificationBlocks` from the new core proof encoder, this is no longer relevant here.

File: vir/defs/polymorphic/ast/stmt.rs (hunk 3 above)
Comment (zgrannan) [reply vir/defs/polymorphic/ast/stmt.rs]: Done!

File: prusti-viper/src/encoder/specification_blocks.rs (hunk 4 above)
Comment (zgrannan) [reply prusti-viper/src/encoder/specification_blocks.rs]: Using it directly now.

**Comment by Aurel300** [comment]
Thanks!

bors r+

**Comment by zgrannan** [comment]
Finished benchmarking commit (9054e05d9fc5bcfbf0576c40c12dfb31a8b3703c): [comparison url](http://3.94.193.1:2346/compare.html?start=aaf4cae0493ca69529b2b53f0590e23126cb83fe&end=9054e05d9fc5bcfbf0576c40c12dfb31a8b3703c&stat=wall-time).

Benchmark returns the average of 3 consecutive runs of the benchmark suite, performed after an initial 'warm-up' run. An external Prusti server is maintained over the 4 runs.
         
### [Wall time](http://3.94.193.1:2346/compare.html?start=aaf4cae0493ca69529b2b53f0590e23126cb83fe&end=9054e05d9fc5bcfbf0576c40c12dfb31a8b3703c&stat=wall-time)
This benchmark run did not return any relevant results for this metric.

### [Instruction count (excludes Viper)](http://3.94.193.1:2346/compare.html?start=aaf4cae0493ca69529b2b53f0590e23126cb83fe&end=9054e05d9fc5bcfbf0576c40c12dfb31a8b3703c&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Cycles (excludes Viper)](http://3.94.193.1:2346/compare.html?start=aaf4cae0493ca69529b2b53f0590e23126cb83fe&end=9054e05d9fc5bcfbf0576c40c12dfb31a8b3703c&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

If you disagree with this performance assessment, please file an issue in [prusti-dev/prusti-perf](https://github.com/prusti-dev/prusti-perf/issues/new).

@rustbot label: -perf-regression

**Pull request merged**